### PR TITLE
New request model

### LINKED ISSUE #82: likeAvirgin has been renamed to reset
as per kuzzleio/kuzzle#559, documentation should be up to date

### PR DESCRIPTION
**Pull request by scottinet**
* update the documentation by replacing the old `RequestObject` and `ResponseObject` with `Request`
* rewrite of parts of the plugins documentation to make it clearer and simpler, especially code examples
* rewrite of controller plugins documentation, using the new, simpler way of writing these plugins (see https://github.com/kuzzleio/kuzzle/pull/563)
* inject @benoitvidis' proposal made in #54 
* fix #82

:warning: events documentation has been left unchanged, since controllers will soon change names

### CODE REVIEW

File: source/includes/plugins/_04-creating-plugins.md [hunk 1]
```diff
@@ -1,340 +1,431 @@
...
-
-POST action:
-
-```
-POST http://kuzzle:7511/api/1.0/_plugin/myplugin/myAction
-{"name": "John Doe"}
-```
-
-- For any other protocol, the _controller name_ attribute is prefixed with the plugin name.
+- HTTP: `GET http://<server>:<port>/<plugin name>/<url defined by the plugin>/<resources>`
```
Comment (dbengsch) [inline source/includes/plugins/_04-creating-plugins.md]: You removed the `_plugin` url part

File: source/includes/plugins/_04-creating-plugins.md [hunk 2]
```diff
@@ -1,340 +1,431 @@
...
+    See the "How plugins receive action arguments" chapter just below
+    for more information.
+   */
+  this.actionImplementation = function (request) {
+    // do action
+    request.setResult(/* action result */);
 
-    // the function must return a Promise:
-    return Promise.resolve(response);
+    return Promise.resolve(request);
```
Comment (dbengsch) [inline source/includes/plugins/_04-creating-plugins.md]: Actually, only the result of the action should be returned, not the request

File: source/includes/plugins/_04-creating-plugins.md [hunk 3]
```diff
@@ -1,340 +1,431 @@
...
+    for more information.
+   */
+  this.otherActionImplementation = function (request) {
+    // do action
+    request.setResult(/* action result */);
 
-  this.init = function (config, pluginContext, isDummy) {
-    this.pluginContext = pluginContext;
-    // do something
+    return Promise.resolve(request);
```
Comment (dbengsch) [inline source/includes/plugins/_04-creating-plugins.md]: same here

**Comment by scottinet** [comment]
@dbengsch > good catches. I fixed other incorrect URLs in the code example too.

File: source/includes/plugins/_04-creating-plugins.md [hunk 4]
```diff
@@ -1,340 +1,440 @@
...
+* modify the provided data on the fly
+* instruct Kuzzle to abort the request, sending an immediate and custom answer to the requesting client
 
-`pipe` plugins functions must take a callback as their last parameter, and this callback must be called at the end of the processing with the following arguments: `callback(error, object)`
+Where a listener exposes a `hooks` property, a pipe plugin needs to expose its events in a `pipes` one, listing the listened events, and the plugin functions to invoke whenever these events are fired (see the example below).
 
-* `error`: if there is an error during the function, this parameter must be set with one of the available Error object provided by the plugin context. Otherwise, set it to `null`
-* `object`: the resulting data, given back to Kuzzle for processing
+Since `pipe` plugins are a part of how Kuzzle process client requests, Kuzzle enforces a timeout on them, rejecting the request altogether if a `pipe` plugin fails to respond in a timely fashion, and forwarding an appropriate [GatewayTimeoutError](#gt-error-gatewaytimeouterror) error to the original client.
```
Comment (xbill82) [inline source/includes/plugins/_04-creating-plugins.md]: "process" :point_right: "processes"

File: source/includes/plugins/_02-plugin-prerequisites.md [hunk 5]
```diff
@@ -1,8 +1,14 @@
 ## Plugin creation prerequisites
 
+### > Writing plugin code
+
+Plugins must be written in Javascript, for the Node.js interpreter. You must be accustomated to [how modules are loaded and exported](https://nodejs.org/docs/v4.6.2/doc/api/modules.html).
```
Comment (scottinet) [inline source/includes/plugins/_02-plugin-prerequisites.md]: Fixed.

File: source/includes/plugins/_04-creating-plugins.md (hunk 4 above)
Comment (scottinet) [inline source/includes/plugins/_04-creating-plugins.md]: Fixed

**Comment by scottinet** [comment]
@xbill82 > are you ok with my corrections? Can we merge this PR?

**Comment by xbill82** [comment]
Yep! Good job! Merged!

**Pull request merged**